prov/efa/src/efa_cq.c

```diff
@@ -0,0 +1,535 @@
+ *
+ * @param[in]		ibv_cq_ex	extended ibv cq
+ * @param[in]		base_ep	    efa_base_ep
+ * @param[in]       cq_entry    fi_cq_tagged_entry
+ */
+static void efa_cq_handle_send_completion(struct efa_base_ep *base_ep,
+					  struct ibv_cq_ex *ibv_cq_ex,
+					  struct fi_cq_tagged_entry *cq_entry)
+{
+	struct util_cq *tx_cq = base_ep->util_ep.tx_cq;
+	int ret = 0;
+
+	/* wr_id = NULL means no completion */
+	if (!ibv_cq_ex->wr_id)
+		return;
+
+	if (base_ep->util_ep.tx_msg_flags == FI_COMPLETION ||
+	    cq_entry->flags & FI_COMPLETION) {
+		/* TX completions should not send peer address to util_cq */
+		if (base_ep->util_ep.caps & FI_SOURCE) {
+			ret = ofi_cq_write_src(tx_cq, cq_entry->op_context,
+					       cq_entry->flags, cq_entry->len,
+					       cq_entry->buf, cq_entry->data,
+					       cq_entry->tag, FI_ADDR_NOTAVAIL);
+		} else {
+			ret = ofi_cq_write(tx_cq, cq_entry->op_context,
+					   cq_entry->flags, cq_entry->len,
+					   cq_entry->buf, cq_entry->data,
+					   cq_entry->tag);
+		}
+
+		if (OFI_UNLIKELY(ret)) {
+			EFA_WARN(FI_LOG_CQ,
+				"Unable to write send completion: %s\n",
+				fi_strerror(-ret));
+			efa_cq_handle_error(base_ep, ibv_cq_ex, tx_cq, -ret,
+					    FI_EFA_ERR_WRITE_SEND_COMP);
+			return;
+		}
+	}
+
+	if (cq_entry->flags != 0)
+		efa_cntr_report_tx_completion(&base_ep->util_ep, cq_entry->flags);
+}
+
+
+/**
+ * @brief handle the event that a receive request has been completed
+ *
+ * @param[in]		ibv_cq_ex	extended ibv cq
+ * @param[in]		base_ep	    efa_base_ep
+ * @param[in]       cq_entry    fi_cq_tagged_entry
+ */
+static void efa_cq_handle_recv_completion(struct efa_base_ep *base_ep,
+					  struct ibv_cq_ex *ibv_cq_ex,
+					  struct fi_cq_tagged_entry *cq_entry)
+{
+	struct util_cq *rx_cq = base_ep->util_ep.rx_cq;
+	fi_addr_t src_addr;
+	int ret = 0;
+	uint64_t cq_flags;
+
+	/* wr_id = NULL means no completion */
+	if (!ibv_cq_ex->wr_id)
+		return;
+
+	cq_flags = (base_ep->util_ep.rx_msg_flags == FI_COMPLETION) ?
+			   0 :
+			   FI_SELECTIVE_COMPLETION;
+
+	if (ofi_need_completion(cq_flags, cq_entry->flags)) {
+		if (base_ep->util_ep.caps & FI_SOURCE) {
+			src_addr = efa_av_reverse_lookup(
+				base_ep->av, ibv_wc_read_slid(ibv_cq_ex),
+				ibv_wc_read_src_qp(ibv_cq_ex));
+			ret = ofi_cq_write_src(rx_cq, cq_entry->op_context,
+					       cq_entry->flags, cq_entry->len,
+					       cq_entry->buf, cq_entry->data,
+					       cq_entry->tag, src_addr);
+		} else {
+			ret = ofi_cq_write(rx_cq, cq_entry->op_context,
+					   cq_entry->flags, cq_entry->len,
+					   cq_entry->buf, cq_entry->data,
+					   cq_entry->tag);
+		}
+
+		if (OFI_UNLIKELY(ret)) {
+			EFA_WARN(FI_LOG_CQ,
+				 "Unable to write recv completion: %s\n",
+				 fi_strerror(-ret));
+			efa_cq_handle_error(base_ep, ibv_cq_ex, rx_cq, -ret,
+					    FI_EFA_ERR_WRITE_SEND_COMP);
+			return;
+		}
+	}
+
+	if (cq_entry->flags != 0)
+		efa_cntr_report_rx_completion(&base_ep->util_ep, cq_entry->flags);
```

Comment:
Same comments, IIRC cntr should always be updated no matter a cq entry is written or not. (Double check man page please)

Reply:
Man page says `Counters record the number of requested operations that have completed.` Does that mean update cntr for no completion(null wr_id) too?

Reply:
The rdm code https://github.com/ofiwg/libfabric/blob/main/prov/efa/src/rdm/efa_rdm_ope.c#L915-L924 returns without updating cntr if unable to write a cq entry

Reply:
Yes, look at the rdm code, it update cntr for inject calls as well where we don't write cqe

Reply:
Looks like rdm does not update cntr if ofi_cq_write fails

Reply:
That is a bug then